core/arch/arm/kernel/ree_fs_ta.c

```diff
@@ -112,6 +118,70 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
 	return res;
 }
 
+static TEE_Result check_update_version(struct shdr_bootstrap_ta *hdr)
+{
+	struct shdr_bootstrap_ta hdr_entry = {0};
+	const struct tee_file_operations *ops = NULL;
+	struct tee_file_handle *fh = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	bool entry_found = false;
+	size_t len = 0;
+	uint32_t i = 0;
+	struct tee_pobj pobj = {
+		.obj_id = (void *)ta_ver_db_obj_id,
+		.obj_id_len = sizeof(ta_ver_db_obj_id)
+	};
+
+#ifdef CFG_RPMB_FS
+	ops = tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE_RPMB);
```

Comment:
I was just thinking of configuration where both `CFG_REE_FS` and `CFG_RPMB_FS` are enabled. In this case `TEE_STORAGE_PRIVATE` defaults to `ree_fs_ops` but we want to rely on `rpmb_fs_ops` for more robustness, that's the reason for using `TEE_STORAGE_PRIVATE_RPMB` specifically.

Reply:
You're concerned that REE FS secure storage might become corrupted in some way?
How likely or unlikely that is I don't know, but if it happens we'll have a big problem nonetheless. If you don't trust the REE FS secure storage, you should not use it at all.
I think it would be more clean to trust that what the platform has selected as secure storage instead of trying to override it. It would also be more consistent with TADB.

Reply:
> You're concerned that REE FS secure storage might become corrupted in some way?

No, I am not concerned about corruption of REE-FS but its about how much rollback protection would REE-FS provide while comparing with RPMB. Have a look [here](https://optee.readthedocs.io/en/latest/architecture/secure_storage.html#globalplatform-trusted-storage-requirement). And its very important aspect for information that we are trying to store here.

> I think it would be more clean to trust that what the platform has selected as secure storage instead of trying to override it.

What if a platform enables both RPMB (for more robustness but limited secure storage space) and REE-FS (provides you with ample secure storage space)?

> It would also be more consistent with TADB.

I guess REE-FS is more suited than RPMB due to storage requirements for large TAs.

Reply:
> > You're concerned that REE FS secure storage might become corrupted in some way?
> 
> No, I am not concerned about corruption of REE-FS but its about how much rollback protection would REE-FS provide while comparing with RPMB. Have a look [here](https://optee.readthedocs.io/en/latest/architecture/secure_storage.html#globalplatform-trusted-storage-requirement). And its very important aspect for information that we are trying to store here.

If RPMB is available it will be used by REE FS secure storage for rollback protection.

> 
> > I think it would be more clean to trust that what the platform has selected as secure storage instead of trying to override it.
> 
> What if a platform enables both RPMB (for more robustness but limited secure storage space) and REE-FS (provides you with ample secure storage space)?


Then REE FS secure storage will have proper rollback protection


> 
> > It would also be more consistent with TADB.
> 
> I guess REE-FS is more suited than RPMB due to storage requirements for large TAs.

The TAs are not stored directly in secure storage, see https://optee.readthedocs.io/en/latest/architecture/trusted_applications.html#secure-storage-ta

Reply:
Ah I see, thanks for this information. I will make it consistent with TADB then.